crates/persistent-storage/src/lib.rs

```diff
@@ -0,0 +1,41 @@
+use base64::{engine::general_purpose, Engine as _};
+use ethers::prelude::*;
+use ethers_signers::WalletError;
+use reqwest::{multipart, Client};
+use serde::Deserialize;
+
+#[derive(Deserialize)]
+#[allow(non_snake_case)]
+struct IpfsCreateResponse {
+    Hash: String,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum IPFSError {
+    #[error("Authentication message signature error: {0:?}")]
+    AuthSigning(#[from] WalletError),
+    #[error("Invalid Peer ID: {0:?}")]
+    Network(#[from] reqwest::Error),
+}
+
+pub async fn write_to_ipfs(client: Client, file: String) -> Result<String, IPFSError> {
```

Comment:
Why do we save `String`, and not `Vec<u8>`? Does data stored in Crust have to be valid UTF-8 text?

Reply:
We're uploading a text file to IPFS so mostly yes. If it's more convinient, we can do conversion Vec<u8> -> String here I guess

Reply:
No I mean, why cannot we upload a binary file? Not every `Vec<u8>` is a valid `String`.